Implement serializable getattr support for tensor subclasses

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145772

builtins.getattr is not serializable, so we replace it with a custom op that has more refined schema. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D68899421](https://our.internmc.facebook.com/intern/diff/D68899421)

tugsbayasgalan committed:
Implement serializable getattr support for tensor subclasses

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with an aten op that has more refined schema. 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with an aten op that has more refined schema. Note that this op only makes sense in python land so we error in C++. 

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -187,6 +187,10 @@
   dispatch:
     CPU: _functional_assert_async_msg_cpu
 
+- func: _access_subclass_inner_tensor(Tensor src, str attr) -> Tensor
```
It's not clear to me that we want this to be a dispatcher op (since it doesn't really need to go through the dispatcher at all). Any reason not to make it a special function that lives somewhere under `torch.export`?

tugsbayasgalan replied:
torch.export has a good ser/der support for aten ops, but i can do the special function way too

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -1528,7 +1527,7 @@ def custom_getattribute(self, attr, *, original_getattr, attrs_to_proxy):
                                     inner_proxy = tracer.create_proxy(
                                         "call_function",
-                                        builtins.getattr,
+                                        torch.ops.aten._access_subclass_inner_tensor.default,
```
hmm I think we also agreed to "const-propagate through any constant metadata" as part of the getattr changes

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with an aten op that has more refined schema. Note that this op only makes sense in python land so we error in C++. 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with an aten op that has more refined schema. Note that this op only makes sense in python land so we error in C++. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with a custom op that has more refined schema. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68899421).

bdhirsh reviewed (approved):
sgtm. Just calling out the plan-on-record again:

(1) consensus from composability was that in the end-state, we should *not* trace these getattrs by default, and instead default to autograd HOPs whenever we see subclass usages in the graph. The major benefit this provides is that subclasses don't have a BC surface in the pre-dispatch IR anymore, although we agreed that it won't work in all cases for training

(2) this is still useful though, since (eventually) we'll need a way for users to turn off the joint-graph-HOP stuff and trace subclass ops directly, in cases where the HOP won't work for them

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with a custom op that has more refined schema. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68899421](https://our.internmc.facebook.com/intern/diff/D68899421)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with a custom op that has more refined schema. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68899421](https://our.internmc.facebook.com/intern/diff/D68899421)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68899421).

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with a custom op that has more refined schema. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68899421](https://our.internmc.facebook.com/intern/diff/D68899421)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68899421).

tugsbayasgalan committed:
Update on "Implement serializable getattr support for tensor subclasses"


builtins.getattr is not serializable, so we replace it with a custom op that has more refined schema. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68899421](https://our.internmc.facebook.com/intern/diff/D68899421)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68899421).
